File: go-algorithm/leetcode/problem337.go
package main

// 337. House Robber III
// https://leetcode.com/problems/house-robber-iii/

func rob3(root *TreeNode) int {
	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}
	memo0 := map[*TreeNode]int{}
	memo1 := map[*TreeNode]int{}
	memo := [2]map[*TreeNode]int{memo0, memo1}
	var dfs func(*TreeNode, int) int
	dfs = func(node *TreeNode, can int) int {
		if node == nil {
			return 0
		}
		if memo[can][node] > 0 {
			return memo[can][node]
		}
		ret := dfs(node.Left, 1) + dfs(node.Right, 1)
		if can > 0 {
			ret = max(ret, node.Val+dfs(node.Left, 0)+dfs(node.Right, 0))
		}
		memo[can][node] = ret
		return ret
	}
	return dfs(root, 1)
}

File: python-algo/leetcode/problem_337.py
"""337. House Robber III
https://leetcode.com/problems/house-robber-iii/
"""
from common.tree_node import TreeNode
from typing import Optional
import functools


class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        @functools.cache
        def dfs(node: Optional[TreeNode], flag: bool) -> int:
            if not node:
                return 0
            ret = dfs(node.left, True) + dfs(node.right, True)
            if flag:
                ret = max(ret, dfs(node.left, False) + dfs(node.right, False) + node.val)
            return ret

        return max(dfs(root, True), dfs(root, False))
